File: 2021/ЛБ2_хтс/project/flow.pas
uses URegularFunctions;

type 
  Flow = class
    mass_flow_rate, mole_flow_rate, volume_flow_rate: real;
    mass_fractions, mole_fractions, volume_fractions: array of real;
    temperature: real;
    density: real;
    molar_mass: real;
    heat_capacity: real;
    
    constructor (mass_flow_rate: real; mass_fractions: array of real; 
                 temperature: real);
    begin
      self.mass_flow_rate := mass_flow_rate;
      self.mass_fractions := normalize(mass_fractions);
      self.temperature := temperature;
      
      self.mole_fractions := convert_mass_to_mole_fractions(self.mass_fractions);
      self.volume_fractions := convert_mass_to_volume_fractions(self.mass_fractions);
      
      self.density := get_flow_density(self.mass_fractions);
      self.molar_mass := get_flow_molar_mass(self.mass_fractions);
      self.heat_capacity := get_heat_capacity(self.mass_fractions, 
                                              self.temperature);
                                              
      self.volume_flow_rate := self.mass_flow_rate / self.density / 1000;
      self.mole_flow_rate := self.mass_flow_rate / self.molar_mass;
    end;             
  
  end;
  

begin
  var mass_frac := ArrRandomReal(24, 1e-5, 0.999);
  var f1 := new Flow(1000, mass_frac, 273.15);
  f1.volume_flow_rate.Println;
  f1.mole_flow_rate.Println;
end.

File: 2024/Lec/ЛК-06-Основы ООП/scripts/flow.py
import numpy as np


class Flow:
    def __init__(self, mass_flow_rate: float,
                 mass_fractions: np.ndarray,
                 temperature: float, pressure: float) -> None:
        
        self.mass_flow_rate = mass_flow_rate
        self.mass_fractions = mass_fractions
        self.temperature = temperature
        self.pressure = pressure

    def convert_mass_to_volume_fractions(self,
                                         densities: np.ndarray) -> np.ndarray:
        x = self.mass_fractions / densities
        s = x.sum()
        return x / s

    def convert_mass_to_mole_fractions(self, mr: np.ndarray) -> np.ndarray:
        x = self.mass_fractions / mr
        s = x.sum()
        return x / s

    def get_flow_density(self, densities: np.ndarray) -> float:
        return (self.mass_fractions / densities).sum() ** -1

    def get_average_molar_mass(self, mr: np.ndarray) -> float:
        return (self.mass_fractions / mr).sum() ** -1


if __name__ == '__main__':
    densities = np.array([.416, .546, .585, .5510, .6, .616, .6262, .6594])
    mr = np.array([16, 30, 44, 58, 58, 72, 72, 86])
    mass_fractions = np.array([.1, .1, .1, .4, .2, .05, .03, .02])

    f1 = Flow(mass_flow_rate=100, mass_fractions=mass_fractions,
              temperature=150, pressure=101.325)

    f1.volume_fractions = f1.convert_mass_to_volume_fractions(densities)
    f1.mole_fractions = f1.convert_mass_to_mole_fractions(mr)
    f1.density = f1.get_flow_density(densities)
    f1.mol_mass = f1.get_average_molar_mass(mr)

    print(f1.volume_fractions)
    print(f1.mole_fractions)
    print(f1.density)
    print(f1.mol_mass)
